core/domain/exp_services.py

```diff
@@ -586,6 +588,277 @@ def apply_change_list(exploration_id, change_list):
+        del states_version_history[state_name]
+
+    # Now, add the states which were newly added during this commit. The
+    # version history of these states are initialized as None because they
+    # were newly added and have no 'previously edited version'.
+    for state_name in exp_versions_diff.added_state_names:
+        states_version_history[state_name] = (
+            state_domain.StateVersionHistory(None, None, committer_id))
+
+    # Now, handle the updation of version history of states which were renamed.
+    # Firstly, we need to clean up the exp_versions_diff.old_to_new_state_names
+    # dict from the state names which are not effectively changed. For example,
+    # if a state was renamed from state_1 to state_2 and then from state_2 to
+    # state_1 in the same commit, then there is no effective change in state
+    # name and we need to clear them from this dict.
+    effective_old_to_new_state_names = {}
+    for old_state_name, new_state_name in (
+        exp_versions_diff.old_to_new_state_names.items()):
+        if old_state_name != new_state_name:
+            effective_old_to_new_state_names[old_state_name] = new_state_name
+    for old_state_name in effective_old_to_new_state_names:
+        del states_version_history[old_state_name]
+    for old_state_name, new_state_name in (
+        effective_old_to_new_state_names.items()):
+        states_version_history[new_state_name] = (
+            state_domain.StateVersionHistory(
+                prev_version, old_state_name, committer_id))
+
+    # The following list includes states which exist in both the old states
+    # and new states and were not renamed.
+    states_which_were_not_renamed = []
+    for state_name in old_states:
+        if (
+            not(state_name in exp_versions_diff.deleted_state_names) and
+            not(state_name in effective_old_to_new_state_names)
+        ):
+            states_which_were_not_renamed.append(state_name)
+
+    # We have dealt with state additions, deletions and renames.
+    # Now we deal with states which were present in both versions and
+    # underwent changes only through the command EDIT_STATE_PROPERTY.
+    # The following dict stores whether the properties of states present
+    # in states_which_were_not_renamed were changed using EDIT_STATE_PROPERTY.
+    state_property_changed_data = {
+        state_name: False
+        for state_name in states_which_were_not_renamed
+    }
+    state_property_ignore_list = [
+        exp_domain.STATE_PROPERTY_RECORDED_VOICEOVERS,
+        exp_domain.STATE_PROPERTY_WRITTEN_TRANSLATIONS
+    ]
+    for change in change_list:
+        if (
+            change.cmd == exp_domain.CMD_EDIT_STATE_PROPERTY and
+            change.property_name not in state_property_ignore_list
+        ):
+            state_name = change.state_name
+            if state_property_changed_data.get(state_name) is False:
+                state_property_changed_data[state_name] = True
```

Comment:
```suggestion
            state_property_changed_data[change.state_name] = True
```

Reply:
We cannot do this here because the `state_name` may not be a key in the dict `state_property_changed_data`. This dict only contains those states which were present in both prev_version and new_version and were not renamed. Hence, we need to check if state_name is present in the dict or not.

This portion of the code handles updation of version history of those states which were only modified by the command EDIT_STATE_PROPERTY and nothing else.

Reply:
Ah okay, then maybe lets do it this way:
```suggestion
            state_name = change.state_name
            if state_name in state_property_changed_data:
                state_property_changed_data[state_name] = True
```

The functionality is the same but it suggests that we actually mostly check whether `state_property_changed_data` contains that state or not. One other possibliity would be:
```suggestion
            state_name = change.state_name
            if state_property_changed_data.get(state_name) is not None:
                state_property_changed_data[state_name] = True
```

but I think the first one is better.

Reply:
Done with the first one.